service/history/workflow/delete_manager.go

```diff
@@ -94,13 +90,26 @@ func (m *DeleteManagerImpl) AddDeleteWorkflowExecutionTask(
 	nsID namespace.ID,
 	we commonpb.WorkflowExecution,
 	ms MutableState,
+	transferQueueAckLevel int64,
+	visibilityQueueAckLevel int64,
 ) error {
 
 	if ms.IsWorkflowExecutionRunning() {
 		// Running workflow cannot be deleted. Close or terminate it first.
 		return consts.ErrWorkflowNotCompleted
 	}
 
+	// Create delete workflow execution task only if workflow is closed successfully and all pending tasks are completed.
+	// Otherwise, mutable state might be deleted before close tasks are executed.
+	// Unfortunately, queue ack levels are updated with delay (default 60s),
+	// therefore this API will return error if workflow is deleted within 60 seconds after close.
+	if (ms.GetExecutionInfo().CloseTransferTaskId != 0 && // backward compatibility (remove in 1.18).
+		ms.GetExecutionInfo().CloseTransferTaskId > transferQueueAckLevel) || // Transfer close task wasn't executed.
+		(ms.GetExecutionInfo().CloseVisibilityTaskId != 0 && // backward compatibility (remove in 1.18).
+			ms.GetExecutionInfo().CloseVisibilityTaskId > visibilityQueueAckLevel) {
+		return consts.ErrWorkflowNotReady
+	}
+
```

Comment:
This will fail the DeleteWorkflow API call. I think it is better to let this task in, and verify / wait this in the task processing. If the ack level is not reached, fail the task processing and let the task processing automatically retry it.

Reply:
Both approaches look reasonable to me. 
With the delete flag, only tasks for recently closed wf will have to wait, so task processing shouldn't be blocked.

Reply:
Also visibility delete task doesn't have access to mutable state and moving this check there will add penalty (db read, which in case of Cassandra will go to the deepest internal SS) to every visibility delete operation.